src/uct/api/uct.h

```diff
@@ -2198,6 +2230,56 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
                            uct_allocated_memory_t *mem);
 
 
+/**
+ * @ingroup UCT_MD
+ * @brief Allocate memory of a specified memory type for zero-copy communications
+ * and remote access.
+ *
+ * Allocate potentially registered memory. Every one of the provided allocation
+ * methods will be used, in turn, to perform the allocation, until one succeeds.
+ * Note that if underlying allocators for the specified memory type cannot honor
+ * the requested method, then an error such as UCS_INVALID_PARAM maybe returned.
+ * Whenever the MD method is encountered, every one of the provided MDs will be
+ * used, in turn, to allocate the memory, until one succeeds, or they are
+ * exhausted. In this case the next allocation method from the initial list will
+ * be attempted.
+ *
+ * @param [in]     addr        If @a addr is NULL, the underlying allocation routine
+ *                             will choose the address at which to create the mapping.
+ *                             If @a addr is non-NULL but UCT_MD_MEM_FLAG_FIXED is
+ *                             not set, the address will be interpreted as a hint
+ *                             as to where to establish the mapping. If @a addr is
+ *                             non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then
+ *                             the specified address is interpreted as a requirement.
+ *                             In this case, if the mapping to the exact address
+ *                             cannot be made, the allocation request fails.
+ * @param [in]     min_length  Minimal size to allocate. The actual size may be
+ *                             larger, for example because of alignment restrictions.
+ * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
+ * @param [in]     methods     Array of memory allocation methods to attempt.
+ * @param [in]     num_methods Length of 'methods' array.
+ * @param [in]     mds         Array of memory domains to attempt to allocate
+ *                             the memory with, for MD allocation method. If
+ *                             none of the provided mds can allocate memory of
+ *                             the specified memory type, then allocation fails
+ *                             if MD allocation method is requested.
+ * @param [in]     num_mds     Length of 'mds' array. May be empty, in such case
+ *                             'mds' may be NULL, and MD allocation method will
+ *                             be skipped.
+ * @param [in]     mem_type    Type of memory to be allocated.
```

Comment:
does it make `methods` parameter irrelevant?
Also maybe define extentable function, using parameters structure - `uct_mem_alloc_param_t`?

Reply:
So far methods would make sense for the case when mem_type = TYPE_HOST. But there are also a subset of such methods available in cuda that could used for future purposes.

Defining an extensible function makes sense. Will iterate on this. Thanks!